metaflow/client/core.py

```diff
@@ -475,6 +475,44 @@ def __contains__(self, id: str):
         """
         return bool(self._get_child(id))
 
+    def __setstate__(self, state):
+        """
+        This function is used during the unpickling operation.
+        More info here https://docs.python.org/3/library/pickle.html#object.__setstate__
+        """
+        if isinstance(state, dict):
+            # For backward compatibility: handles pickled objects that were serialized as a `dict`
+            self.__init__(
+                pathspec=state.get("_pathspec", None),
+                attempt=state.get("_attempt", None),
+                _namespace_check=state.get("_namespace_check", None),
+            )
+        elif isinstance(state, tuple):
```

Comment:
to simplify our lives a little bit, I would actually pickle something like this:
```
{
  "version": 1,
  "data": [self.pathspec, self._attempt, self._namespace_check]
}
```

This would replace the check on the type (and instead you can check if state contains "version" -- it won't unless it's the new format). You now have two cases for now:
  - no version (it's the case you have above)
  - a version and you can write a "unpickle_v1" function that does what you do below passing it data

If we ever want to change it again, it will be easy to just add a new "unpickle_v2" function and maintain backward compatibility that way. It's a bit more data in terms of number of bytes for the object but I don't think it makes that big of a difference and makes it a little easier to reason about going forward.

As an example, if we had had this, we would have changed to version 2 and added an additional field easily that way.

Reply:
Actually, we can also maybe do something like this: `"version": "2.8.4"` (or whatever but basically the version of Metaflow from which we have this format). That way, we could give a very nice error message like: "Unpickling this object requires Metaflow version 2.8.4 or greater". Newer versions would have the proper if/else clauses and the default else clause would print out this message.